stages/org.osbuild.luks2.format

```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/python3
+"""
+Create an LUKS2 container via `cryptsetup`.
+
+This stage formats the given `device` to be a Linux Unified Key Setup,
+LUKS version 2, container and set the key to be `passphrase`. The uuid
+of the container must be specified via the coressponding option.
+
+Use the corresponding `org.osbuild.luks2` device to open the container
+during build.
+
+Buildhost commands used: `cryptsetup`.
+"""
+
+
+import os
+import subprocess
+import sys
+
+
+import osbuild.api
+
+
+SCHEMA_2 = r"""
+"devices": {
+  "type": "object",
+  "additionalProperties": true,
+  "required": ["device"],
+  "properties": {
+    "device": {
+      "type": "object",
+      "additionalProperties": true
+    }
+  }
+},
+"options": {
+  "additionalProperties": false,
+  "required": ["passphrase", "uuid"],
+  "properties": {
+    "passphrase": {
+        "description": "Passphrase to use",
+        "type": "string"
+    },
+    "uuid": {
+      "description": "UUID for the LUKS device to use",
+      "type": "string"
+    },
+    "cipher": {
+      "description": "Cipher to use",
+      "type": "string"
+    },
+    "label": {
+      "description": "Label to use",
+      "type": "string"
+    },
+    "subsystem": {
+      "description": "Additional label",
+      "type": "string"
+    },
+    "sector-size": {
+      "description": "Sector size to use",
+      "type": "integer"
+    }
+  }
+}
```

Comment:
for edge, adding `--pbkdf-memory` also allows us to set that regardless of how much ram the build machine has (i.e., we're always gonna set it to 512mb)

Reply:
What about the other options like `--pbkdf-force-iterations`, `--pbkdf-parallel`?

Reply:
I don't think we would need to set that.
In the worst case, it means that a device is busy slightly longer when unlocking the file system (but that's still in the order of seconds), but it shouldn't prevent unlocking the disk fully.
So I think that for force-iterations and parallize parameters, we should just keep using the defaults.

Reply:
If I understand the man page correctly (currently reading the source to make sure) it actually does a benchmark on the fly to figure out the iterations, so that would imply the build host having an impact on the result.

Reply:
Had a quick look into the cryptsetup source, and if you supply `force-iterations` it will set `CRYPT_PBKDF_NO_BENCHMARK` (https://gitlab.com/cryptsetup/cryptsetup/-/blob/master/src/utils_luks.c#L130) but otherwise `iterations` will be `0` and that will trigger a benchmark:
```
	/* Reset iteration count so benchmark must run again. */
	if (cd_pbkdf->flags & CRYPT_PBKDF_NO_BENCHMARK)
		cd_pbkdf->iterations = pbkdf->iterations;
	else
		cd_pbkdf->iterations = 0;
```

I think that is suboptimal, since we would do that for every image build and it will also vary from build to build and depend on the worker. I think we should rather pick something that makes sense for a range of devices and if that is not easily possible maybe we need some sort of high level blueprint parameter at the composer level?

Btw, the minimal parameters for are defined here https://gitlab.com/cryptsetup/cryptsetup/-/blob/master/lib/crypto_backend/pbkdf_check.c#L42

Reply:
That sounds like a decent approach, sure, but I do think that if we decide to set these parameters by hand, we need to talk with experts about which values to use.

But yeah, my previous reply basically was me saying that yes, we would be benchmarking, and the iterations would be tuned for the (possibly faster) builder system, but that the impact of higher iteration counts (longer decryption times) is higher than the memory parameter (total inability to decrypt with less memory than 1/4th the builders').

Reply:
So, I thought a bit more about this and actually, I think this is something that should probably be done at installation time, i.e. when we do create the new key that is bound to the TPM.
The `luksAddKey` and `luksChangeKey` for `cryptsetup` both have all the `pbkdf` options, which makes sense. So, when building, using the null_cipher and a blank passphrase we might as well specify the minimum and then adjust the value during the new key enrolment (maybe clevis is already doing that anyway, or is using the defaults).

Reply:
Regardless of how these things end up (I like @gicmo's suggestion of doing it at install time), we should not rely on the build host for anything, so having it as a parameter in the manifest makes sense.

Reply:
Added the parameter and made it mandatory. I used the minimum values in the sample image and it reduced the image build time, mostly because it does not do the benchmarking anymore (and of course because I used the minimum values).